index.d.ts

```diff
@@ -0,0 +1,104 @@
+import stream = require('stream');
+
+declare function request(superagent: any, promise: any): request.SuperAgentStatic;
+
+declare namespace request {
+  interface SuperAgentStatic extends SuperAgent<SuperAgentRequest> {
+    (url: string): SuperAgentRequest;
+    (method: string, url: string): SuperAgentRequest;
+
+    agent(): SuperAgent<SuperAgentRequest>;
+  }
+
+  interface SuperAgent<Req extends Request<any>> extends stream.Stream {
+    get(url: string): Req;
+    post(url: string): Req;
+    put(url: string): Req;
+    head(url: string): Req;
+    del(url: string): Req;
+    delete(url: string): Req;
+    options(url: string): Req;
+    trace(url: string): Req;
+    copy(url: string): Req;
+    lock(url: string): Req;
+    mkcol(url: string): Req;
+    move(url: string): Req;
+    purge(url: string): Req;
+    propfind(url: string): Req;
+    proppatch(url: string): Req;
+    unlock(url: string): Req;
+    report(url: string): Req;
+    mkactivity(url: string): Req;
+    checkout(url: string): Req;
+    merge(url: string): Req;
+    notify(url: string): Req;
+    subscribe(url: string): Req;
+    unsubscribe(url: string): Req;
+    patch(url: string): Req;
+    search(url: string): Req;
+    connect(url: string): Req;
+
+    parse(fn: Function): Req;
+    saveCookies(res: Response): void;
+    attachCookies(req: Req): void;
+  }
+
+  interface Response extends NodeJS.ReadableStream {
+    text: string;
+    body: any;
+    files: any;
+    header: any;
+    type: string;
+    charset: string;
+    status: number;
+    statusType: number;
+    info: boolean;
+    ok: boolean;
+    redirect: boolean;
+    clientError: boolean;
+    serverError: boolean;
+    error: Error;
+    accepted: boolean;
+    noContent: boolean;
+    badRequest: boolean;
+    unauthorized: boolean;
+    notAcceptable: boolean;
+    notFound: boolean;
+    forbidden: boolean;
+    get(header: string): string;
+  }
+
+  interface Request<Req extends Request<any>> /* extends NodeJS.WritableStream */ {
```

Comment:
why do we need a type parameter here? As far as I understand all the methods simply return `this`